### PR TITLE
Add test_feder

### PR DESCRIPTION
**Pull request by cydrain**

### CODE REVIEW

File: tests/ut/test_feder.cc [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+        REQUIRE(id >= 0);
+        REQUIRE(id < nb);
+    }
+}
+
+void
+CheckHnswVisitInfo(const knowhere::DataSetPtr result, int64_t nb) {
+    auto json_info = result->GetJsonInfo();
+    auto json_id_set = result->GetJsonIdSet();
+    // std::cout << json_info << std::endl;
```
Comment (zhengbuqian) [inline tests/ut/test_feder.cc:70]: ditto

File: tests/ut/test_feder.cc [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+#include "knowhere/feder/IVFFlat.h"
+#include "knowhere/log.h"
+#include "utils.h"
+
+void
+CheckHnswMeta(const knowhere::DataSetPtr result, int64_t nb, const knowhere::Json& cfg) {
+    auto json_info = result->GetJsonInfo();
+    auto json_id_set = result->GetJsonIdSet();
+    // std::cout << json_info << std::endl;
+    std::cout << "json_info size = " << json_info.size() << std::endl;
```
Comment (zhengbuqian) [inline tests/ut/test_feder.cc:31]: Consider using INFO macros provided by catch2?

File: tests/ut/test_feder.cc [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    for (auto& cluster : clusters) {
+        for (auto id : cluster.node_ids_) {
+            REQUIRE(id >= 0);
+            REQUIRE(id < nb);
+            all_id_set.insert(id);
+        }
+    }
+    REQUIRE(all_id_set.size() == (size_t)nb);
+
+    // check IDSet
```
Comment (zhengbuqian) [inline tests/ut/test_feder.cc:134]: What do we check here? Are we only expecting the json operations to succeed? pls comment if so.

File: tests/ut/test_feder.cc [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    };
+
+    auto load_raw_data = [](knowhere::Index<knowhere::IndexNode>& index, const knowhere::DataSet& dataset,
+                            const knowhere::Json& conf) {
+        auto rows = dataset.GetRows();
+        auto dim = dataset.GetDim();
+        auto p_data = dataset.GetTensor();
+        knowhere::BinarySet bs;
+        auto res = index.Serialize(bs);
+        REQUIRE(res == knowhere::Status::success);
```
Comment (zhengbuqian) [inline tests/ut/test_feder.cc:180]: `REQUIRE(index.Serialize(bs) == knowhere::Status::success);` and remove `res`

File: tests/ut/test_feder.cc [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+    auto hnsw_gen = [&base_gen]() {
+        knowhere::Json json = base_gen();
+        json[knowhere::indexparam::HNSW_M] = 8;
+        json[knowhere::indexparam::EFCONSTRUCTION] = 200;
+        json[knowhere::indexparam::EF] = 16;
+        json[knowhere::indexparam::OVERVIEW_LEVELS] = 2;
+        return json;
+    };
+
+    auto load_raw_data = [](knowhere::Index<knowhere::IndexNode>& index, const knowhere::DataSet& dataset,
```
Comment (zhengbuqian) [inline tests/ut/test_feder.cc:173]: why make this chunk lambda if only called once in "Test IVF_FLAT Feder"?

**Review by zhengbuqian — COMMENTED** [review]
/lgtm

with minor comments

File: tests/ut/test_feder.cc (hunk 1 above)
Comment (cydrain) [reply tests/ut/test_feder.cc:70]: removed

File: tests/ut/test_feder.cc (hunk 2 above)
Comment (cydrain) [reply tests/ut/test_feder.cc:31]: replaced with LOG_KNOWHERE_INFO

File: tests/ut/test_feder.cc (hunk 3 above)
Comment (cydrain) [reply tests/ut/test_feder.cc:134]: update

File: tests/ut/test_feder.cc [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+#include "knowhere/feder/HNSW.h"
+#include "knowhere/feder/IVFFlat.h"
+#include "knowhere/log.h"
+#include "utils.h"
+
+void
+CheckHnswMeta(const knowhere::DataSetPtr result, int64_t nb, const knowhere::Json& cfg) {
+    auto json_info = result->GetJsonInfo();
+    auto json_id_set = result->GetJsonIdSet();
+    // std::cout << json_info << std::endl;
```
[the remark this answers is not in the document]
Comment (cydrain) [reply tests/ut/test_feder.cc]: removed

**Pull request merged**